### ISSUE TITLE
Can't make relative dates work

### ISSUE THREAD

**Issue by slorber** [yahoo/react-intl#84]
Hi,

I don't really know why but when trying to use relative format for dates I get errors.

I've tried to debug this and what I can see is that:

`function MessageFormat(message, locales, formats)`

This function is called with message = "en-EN", and locale being an empty object.

Then it fails because of "undefined is not a function" on 

```
    throw new Error(
        'No locale data has been added to IntlMessageFormat for: ' +
        locales.join(', ')
    );
```

I use version 1.1.0 with relativeformat 1.0.3

Any idea?

**Comment by caridy** [comment]
What is `en-EN`? Normally, you have `en-<CountryCode>`, or just `en`. But even if that locale can't be found (or does not exists), it should have fallback to `en`, which is always present. We will investigate.

**Comment by slorber** [comment]
Hmm it's a mistake because I'm french, I meant `en-UK` or `en-US` or `fr-FR` but actually this does not work either

**Comment by caridy** [comment]
Well, according to the error, it seems that you're not providing `locales`, which is suppose to be an array. Make sure you're passing the right locales at the top level of your app, when rendering the top level react component, make sure you also add the mixin into that component so the locales can be propagated all the way down to the child components.

**Comment by slorber** [comment]
I am sure I did actually. 

In Chrome I see in the call stack:

```
case 'relative':
                return this.getRelativeFormat(locales, options).format(value);
```

Variables are:
```
locales: "fr"
options: {}
this: ReactCompositeComponent.createClass.Constructor
type: "relative"
value: Fri Feb 27 2015 14:23:43 GMT+0100 (CET)
```

**Comment by caridy** [comment]
are you loading `fr` locale data? since locales is `fr`, it will try to access the data for that one.

**Comment by slorber** [comment]
I think so. But as I'm using Browserify and FormatJs seems to require a global value to register extra locale files, I may have done an integration that is not very elegant. Is there a way to check the fr locale data is available? 

By the way, I think I tried with simply "en" and it did not work either.

Will check again tomorrow.

**Comment by slorber** [comment]
I can see that __addLocalData is called for both "en" (by default) and "fr":

```
__addLocaleData Object {locale: "en", pluralRuleFunction: function, fields: Object} react-intl.js:5
test adding FR fr.js:1
__addLocaleData Object {locale: "fr", pluralRuleFunction: function, fields: Object} 
```

Here's my integration:

```
var ReactIntl = require('react-intl');
// Required because local data files assume global variable
window.ReactIntl = ReactIntl;
// "en" is provided by default
require("react-intl/dist/locale-data/fr.js");
```

It is a bit hacky but it seems locale data were not meant to be loaded through Browserify

**Comment by caridy** [comment]
ah, yeah, this is a duplicated of https://github.com/yahoo/react-intl/issues/82, let's use that one to follow up on this request.

**Comment by slorber** [comment]
No @caridy this is not the same problem. I have successfully loaded both "fr" and "en" using the above trick. But then it still does not work.

I'm sure it's a bug in the lib and not related to Browserify/CommonJS according to my debug informations.


```javascript
function createFormatCache(FormatConstructor) {
    var cache = src$es5$$.objCreate(null);

    return function () {
        var args    = Array.prototype.slice.call(arguments);
        var cacheId = getCacheId(args);
        var format  = cacheId && cache[cacheId];

        if (!format) {
            format = src$es5$$.objCreate(FormatConstructor.prototype);
            FormatConstructor.apply(format, args);

            if (cacheId) {
                cache[cacheId] = format;
            }
        }

        return format;
    };
}
```

The problem seems to be there.

```
FormatConstructor = MessageFormat(message, locales, formats)

args = args: Array[2]
0: "en"
1: Object

cacheId: "["en",[]]"
```

This code is calling `FormatConstructor` with 2 args instead of 3. The `createFormatCache` is not provided any "message".

I don't know at all why this happens, but actually when `this.getRelativeFormat(locales, options)` is called, it tries to load from the cache a `MessageFormat` instead of a `RelativeFormat`

**Comment by ericf** [comment]
@slorber Can you provide a runnable simplified example project that exhibits this problem? We haven't seen this before in our tests, website, or from other users, and having a full example will allow us to get to the bottom of this.

**Comment by slorber** [comment]
@ericf @caridy I've created a sandbox to reproduce that bug.

https://github.com/slorber/react-intl-relativeformat-bug

To run the code:

- npm install
- gulp
- open build/index.html

I've let all my normal build / integration code of react-intl, and all the libs. This reflects what is done in my real app and permits to reproduce the bug.

**Comment by slorber** [comment]
Note my Browserify Integration is pretty similar to the integration you have done for tests:
https://github.com/yahoo/react-intl/blob/master/tests/browserify/app.js

```
window.ReactIntl = require('../../');
require('../../lib/locales.js');

window.React = require('react');
```

VS

```
var ReactIntl = require('react-intl');
// Required because local data files assume global variable
window.ReactIntl = ReactIntl;
// "en" is provided by default
require("react-intl/dist/locale-data/fr.js");
```

**Comment by slorber** [comment]
@ericf @caridy have you been able to reproduce my bug?

**Comment by ericf** [comment]
@slorber I haven't tried yet; but I'm working on getting a new `intl-relativeformat` release out asap. So before I get that release out I'll make sure to check out your example app to be sure if any bugs exist at the lower-level they are fixed.

**Comment by ericf** [comment]
@slorber This looks like an issue with your build. When looking in the inspector I could see that for some reason the `IntlMessageFormat` constructor is returned when I did:

```js
var IRF = require('react-intl/node_modules/intl-relativeformat');
```

But when doing this in the project's dir in the Node REPL, I got back the `IntlRelativeFormat` constructor as expected. I also double-checked whether or not this was a Browserify issue, but our Browserify tests work correctly for `<FormattedRelative>`.

Something isn't bundling the deps right in your build. You might want to search around to see if this is a common problem. One thing to note is that `intl-relativeformat` depends on `intl-messageformat`, and `react-intl` depends on both of those. My guess would be that this dep graph is confusing your build somewhere…
